src/enumerable.cr

```diff
@@ -1912,9 +1912,20 @@ module Enumerable(T)
     end
   end
 
-  # :nodoc:
-  def first_internal
-    # overridden in Tuple to disable literal index lookup
-    first
+  # Returns a value with the same type as an element of *x*, even if *x* is not
+  # an `Enumerable`.
+  #
+  # Used by splat expansion inside array literals. For example, this code
+  #
+  # ```
+  # [1, *{2, 3.5}, 4]
+  # ```
+  #
+  # will end up calling `typeof(1, ::Enumerable.element_type({2, 3.5}), 4)`.
+  #
+  # NOTE: there should never be a need to call this method outside the standard library.
+  def self.element_type(x)
```

Comment:
This method returns a value, not a type. So maybe it should rather be called `first_element`?
I'd also suggest a more expressive name for the argument. IMO it could be named `enumerable` (even if it does not need to implement `Enumerable`).

Reply:
`*_type` is already used in other places in the standard library (e.g. `Iterator::Flatten.iterator_type`), I don't think this PR needs to adopt a different naming scheme.